### PR TITLE
Circle ci test

### LINKED ISSUE #2: Save Image Link
#### As a user I want to save a link to an image.

_Assignees_: @HaominZhang, @XanderStrike

__Story__:

Now that we have an awesome landing page for our users to experience it is time
to introduce our first feature! In general, we want to deliver value to our
customers as fast as possible. As a result, we have decided that we want our
users to be able to save links to images, and immediately see that we have
saved the URL to the image.

One could argue that simply saving the image, and not showing it, provides some
value. Yes, you could split this story here. However, because saving image URLs
and viewing those images are both relatively small additions that work well
together, it is acceptable to complete them together. Note, on the other hand,
that we do not yet want to implement the image index. Its addition adds too
much complexity, so we’ll save that for a future story.

In general, feature addition, or stories, should be split into the smallest
deliverable chunk that provides value to your users with the caveat that it
might make sense to combine two, sometimes more, tiny and related features into
a single story.

__Acceptance criteria__:
- [ ] The landing page links to the image link submission form.
- [ ] The link is entered through a form.
- [ ] After the form is saved, the link is persisted in the database.
- [ ] After the form is submitted, the user is redirected to a page displaying
  the image.
- [ ] I cannot successfully save an image with an invalid URL.
- [ ] An error message is associated with the appropriate input field on
  failure.

__Discussion Topic__:
- [ ] What is a valid image URL?

__Dependencies__:
- Deployment Environment

__Note__: Many image sharing sites will permit one to upload images, or
download images when provided a link. For this project we will only work with
the link, i.e., URL, to various images.

__Note__: To understand the database tables created by the `acts_as_taggable` gem, you may find it useful to use a program like the [DB Browser for SQLite](http://sqlitebrowser.org/).

### PR DESCRIPTION
**Pull request by ChaoHuangAtAppfolio**
Closes #2 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/fixtures/images.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (darreneng) [inline test/fixtures/images.yml]: If you're not planning on using fixtures in your tests then we can probably just delete this file.

File: test/models/image_test.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (darreneng) [inline test/models/image_test.rb]: Could we remove these commented lines?

File: test/models/image_test.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  #
+  #   # make sure valid? returns false
+  #
+  #   # atomic commits
+  # end
+
+  test 'invalid url' do
+    # below is an invalid url
+    params = { url: '123.adaf' }
+    image = Image.new(params)
```
Comment (darreneng) [inline test/models/image_test.rb]: No need to make a change here, but you can actually pass the params directly to `Image.new` as keyword arguments. So you could write something like `Image.new(url: '123.adaf')`

File: db/migrate/20180522204251_create_images.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class CreateImages < ActiveRecord::Migration[5.1]
+  def change
+    create_table :images do |t|
+      # t.integer :index
```
[the remark this answers is not in the document]
Comment (ChaoHuangAtAppfolio) [reply db/migrate/20180522204251_create_images.rb]: Already deleted.

File: test/models/image_test.rb (hunk 2 above)
Comment (ChaoHuangAtAppfolio) [reply test/models/image_test.rb]: Done.

File: test/models/image_test.rb (hunk 3 above)
Comment (ChaoHuangAtAppfolio) [reply test/models/image_test.rb]: Modified.

File: app/controllers/images_controller.rb [hunk 5]
```diff
@@ -66,4 +58,14 @@ def set_image
   def image_params
     params.require(:image).permit(:index, :url)
   end
+
+  private
+
+  def duplicate?(url)
+    @images = Image.all
+    @images.each do |image|
```
Comment (darreneng) [inline app/controllers/images_controller.rb]: No need to make a change here, but there should be an any function that let you write something more concise. 

https://ruby-doc.org/core-2.2.0/Enumerable.html#method-i-any-3F

**Pull request merged**